File: DS-MADE-EASY/LinkedList/reverseLinkedList.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>
#include<limits.h>
/*

# Author			: @RAJ009F
# Topic or Type 	: LinkedLists 
# Problem Statement	: reverse LL
# Description		:
# Complexity		:
=======================
#sample output
----------------------

=======================
*/

struct SLL
{
	int data;
	struct SLL *next;
};

struct SLL *newNode(int data)
{
	struct SLL *node = (struct SLL *)malloc(sizeof(struct SLL));
	
	if(!node)
			return;
		
		node->next = NULL;
		node->data = data;
		
		return node;
	
	
}

void printList(struct SLL *h)
{
	while(h)
	{
		printf("[%d]->",h->data);
		h=h->next;
	}
	printf("NULL\n");
	
	
}

void reverseSLL(struct SLL **h)
{
	if(!*h || !(*h)->next)
		return;
	struct SLL *p,*q=NULL, *curr=*h;
	while(curr)
	{
		p = curr;
		curr = curr ->next;
		p->next = q;
		q =p;
		
		
		
	}
	*h = p;
	
	
	
}

void reverseRSLL(struct SLL *curr, struct SLL **head)
{
	if(!curr)
		return ;
	if(!(curr->next))
	{
		*head =curr;
		return;
		
	}
	reverseRSLL(curr->next, head);
	curr->next->next = curr;
	curr ->next = NULL;
	
	
	
}
int main()
	{
		struct SLL *head = newNode(1);
		head -> next = newNode(2);
		head->next->next = newNode(3);
		head->next->next->next = newNode(4);
		struct SLL *temp=head->next->next->next->next = newNode(5);
		head->next->next ->next ->next ->next = newNode(6);
		head->next->next->next->next->next->next = newNode(7);
		head->next->next->next->next->next->next->next = newNode(8);
		head->next->next->next->next->next->next->next->next = newNode(9);
		struct  SLL *last=head->next->next->next->next->next->next ->next ->next ->next = newNode(10);
		
		printf("list before reverse:\n");
		printList(head);
		reverseSLL(&head);
		printf("SLL after reverse:\n");
		printList(head);
		reverseRSLL(head,&head);
		printf("SLL after Re-reverse:\n");
		printList(head);
	
	return 0;
	}

File: GFG/LinkedList/ReverseLinkedList.py
# import java.io.*;
# import java.lang.*;
# import java.util.*;

#!/usr/bin/python -O

# // #include <stdio.h>
# // #include <stdlib.h>
# // #include <string.h>
# // #include<stdbool.h>
# // #include<limits.h>

# // #include<iostream>
# // #include<algorithm>
# // #include<string>
# // #include<vector>
# //using namespace std;

# /*

# Author			: @RAJ009F
# Topic or Type 	: GFG/LinkedList
# Problem Statement	: Reverse Linked List ,  both iterative and recursive
# Description		: 
# Complexity		: 
# =======================
# sample output
# ----------------------

# =======================
# */


class Node:
	def __init__(self, data):
		self.data = data
		self.next = None

class LinkedList:
	def __init__(self):
		self.head = None
		
	def push(self,data):
		node =  Node(data)
		node.next =  self.head
		self.head = node
		
	def print_list(self):
		cur =  self.head
		while cur:
			print (cur.data,  )
			cur =  cur.next
	def reverse_list(self):
		prev = None
		cur = self.head
		temp = None
		while cur != None:
			temp = prev
			prev = cur
			cur = cur.next
			prev.next = temp
		self.head = prev
	
	def reverse_list_rec_util(self,head):
		if not head: 
			return
		if not head.next:	
			 self.head = head
			 return
		self.reverse_list_rec_util(head.next)
		head.next.next = head
		head.next = None
		# temp = cur.next
		# cur.next = prev 
		# self.reverse_list_rec_util(temp, cur)
		#return self.head
	def reverse_list_rec(self):
		self.reverse_list_rec_util(self.head)
		pass
	def insert_node(self, key):
		if not self.head :
			self.head = Node(key)
		elif self.head.data>key:
			node = Node(key)
			node.next = self.head
			self.head = node
		else:
			prev = self.head
			cur =  self.head
			
			while cur and cur.data<key:
				prev =  cur
				cur =  cur.next
			node =  Node(key)
			prev.next =  node
			node.next = cur 
				
			
		

if __name__ == "__main__":
	ll = LinkedList();
	ll.push(1)
	ll.push(2)
	ll.push(3)
	ll.push(4)
	ll.push(6)
	ll.push(7)
	ll.push(8)
	ll.push(10)
	#ll.print_list()
	print ("reverse List")
	ll.reverse_list()
	#ll.print_list()
	
	ll.reverse_list_rec()
	print ("reverse list using rrecursive")
	ll.reverse_list_rec()
	ll.print_list()
	ll.insert_node(9)
	ll.print_list()
		
	
	
		
		
		